ruby/lib/google/protobuf/ffi/field_descriptor.rb

```diff
@@ -0,0 +1,326 @@
+      def has?(msg)
+        if msg.class.descriptor != Google::Protobuf::FFI.get_containing_message_def(self)
+          raise TypeError.new "has method called on wrong message type"
+        end
+        unless has_presence?
+          raise ArgumentError.new "does not track presence"
+        end
+
+        Google::Protobuf::FFI.get_message_has msg.instance_variable_get(:@msg), self
+      end
+
+      ##
+      # Tests if this field tracks presence.
+      #
+      # @return [Boolean] True iff this field tracks presence
+      def has_presence?
+        @has_presence ||= Google::Protobuf::FFI.get_has_presence(self)
+      end
+
+      # @param msg [Google::Protobuf::Message]
+      def clear(msg)
+        if msg.class.descriptor != Google::Protobuf::FFI.get_containing_message_def(self)
+          raise TypeError.new "clear method called on wrong message type"
+        end
+        Google::Protobuf::FFI.clear_message_field msg.instance_variable_get(:@msg), self
+        nil
+      end
+
+      ##
+      # call-seq:
+      #     FieldDescriptor.set(message, value)
+      #
+      # Sets the value corresponding to this field to the given value on the given
+      # message. Raises an exception if message is of the wrong type. Performs the
+      # ordinary type-checks for field setting.
+      #
+      # @param msg [Google::Protobuf::Message]
+      # @param value [Object]
+      def set(msg, value)
+        if msg.class.descriptor != Google::Protobuf::FFI.get_containing_message_def(self)
+          raise TypeError.new "set method called on wrong message type"
+        end
+        unless set_value_on_message value, msg.instance_variable_get(:@msg), msg.instance_variable_get(:@arena)
+          raise RuntimeError.new "allocation failed"
+        end
+        nil
+      end
+
+      def map?
+        @map ||= Google::Protobuf::FFI.is_map self
+      end
+
+      def repeated?
+        @repeated ||= Google::Protobuf::FFI.is_repeated self
+      end
+
+      def sub_message?
+        @sub_message ||= Google::Protobuf::FFI.is_sub_message self
+      end
+
+      def wrapper?
+        if defined? @wrapper
+          @wrapper
+        else
+          message_descriptor = Google::Protobuf::FFI.get_subtype_as_message(self)
+          @wrapper = message_descriptor.nil? ? false : message_descriptor.send(:wrapper?)
+        end
+      end
+
+      private
+
+      def initialize(field_def, descriptor_pool)
+        @field_def = field_def
+        @descriptor_pool = descriptor_pool
+      end
+
+      def self.private_constructor(field_def, descriptor_pool)
+        instance = allocate
+        instance.send(:initialize, field_def, descriptor_pool)
+        instance
+      end
+
+      # TODO(jatl) Can this be added to the public API?
+      def real_containing_oneof
+        @real_containing_oneof ||= Google::Protobuf::FFI.real_containing_oneof self
+      end
+
+      # Implementation details below are subject to breaking changes without
```

Comment:
I don't think a comment will be enough to communicate this. Perhaps a name that starts with `__internal` or similar?

Reply:
Move this up to next to the `private` keyword as it was originally intended.